### DISCUSSION TITLE
skynet能不能提供一个服务之间共享的hash表接口

### DISCUSSION THREAD

**Discussion by yuanfengyun** [cloudwu/skynet#1585] [General]
游戏编程中，配置文件通过id去索引内容的场景是绝大多数。简单提供string --> string的hash表，然后带上版本。这样消耗比现有要小很多。现在我们的配置文件比较多，用了sharedata，生成了大量的协程。感觉这样消耗太大了。

**Comment by aasdpiao** [comment]
datacenter 不知道是不是可以满足你的要求
https://github.com/cloudwu/skynet/wiki/ShareData

**Reply by yuanfengyun** [reply]
现有的办法消耗太大了。

**Reply by yuanfengyun** [reply]
简单的一个无锁hash表更实用一些

**Reply by aasdpiao** [reply]
感觉你的标题跟你的提问好像说的不是同一个东西. 配置文件现阶段最优解应该还是sharedata , 一般初始化的时候直接持有一个sharetable.query("xxx")  不是每次获取配置都去query应该问题不大吧. 如果还是太多.就把所有的配置文件用一个key来存.这样应该是最省的.

**Reply by aasdpiao** [reply]
```lua
--[[

config = {

  item_config = {xxxxx},

  xxx_config = {xxxxxx}

}

]]

local config = sharetable.query("cofig")



function get_item_config(xxx)

  return config[item_config][xxx]

end



function get_xxx_config(xxx)

  return config[xxx_config][xxx]

end

```

**Comment by JieTrancender** [comment]
服务初始化时获取所有表的指针，然后缓存着，可能是你需要的。
https://github.com/cloudwu/skynet/blob/448fe14b9efa2b752a8fc27891c0bf43b09ee733/lualib/skynet/sharetable.lua#L235-L249

**Reply by yuanfengyun** [reply]
这个消耗比简单的hash表大太多了。

**Reply by yuanfengyun** [reply]
sharetable暴力遍历更新的手法也不太认同。

**Reply by fanyh** [reply]
我觉得这个事情；可以自行实现1个c服务来代替sharedata/sharetable/datecenter
从根源上来说；skynet内部对load cache的支持是可以选择性支持的

**Reply by yuanfengyun** [reply]
现有的共享方式为了维持TABLE的形式，太暴力了。通过简化数据为一个hash表，会容易很多。

**Reply by yuanfengyun** [reply]
如果几百或者几千个服务同时更新一个大表，这不得瞬间把服务器干翻啊。

**Comment by zhangshiqian1214** [comment]
感觉你要的是这个,这个是带版本的共享内存
```
STM (Software transactional memory) 模块同样基于共享内存，所以也只能用于同一个 skynet 节点内。它是一个试验性模块，不一定比消息传递的方式更好。只是提供一个新思路来进行同一节点内的服务间数据交换。

因为它不经过 skynet 的消息投递，信息传递的时效性比消息投递要好一些。但由于依旧需要在不同的 lua vm 间交换数据，序列化（使用 skynet.pack 和 skynet.unpack）必不可少。因为绕过了消息投递，它还可以用于广播。多个读取者可以同时去读一个写入者更新的数据。
```

**Comment by ylme** [comment]
> 戏编程中，配置文件通过id去索引内容的场景是绝大多数。简单提供string --> string的hash表，然后带上版本。这样消耗比现有要小很多。现在我们的配置文件比较多，用了sharedata，生成了大量的协程。感觉这样消耗太大了。

用 sharedata 的话，建议在服务端把所有配置文件都读到一个 share data 对象中去。每个文件一个 sharedata 对象不显示，毕竟一个 sharedata 对象会用到一个 lua vm 。
